### ISSUE TITLE
Retrieve node from nodeId

### ISSUE THREAD

**Issue by rbairwell** [ChromeDevTools/devtools-protocol#20]
At the moment, there appears to be no way of actually getting a Node element (including the nodeType, nodeName etc) from a NodeId in the DOM. Most things in DOM appear to return a nodeId, but to actually get the Node from a matching NodeId, you need to do a DOM.getDocument, DOM.requestChildNodes and then walk through the DOM "manually".

Would it be possible to add a DOM.getNodeById({nodeId:[nodeId]}) method which returns the relevant node? (I did originally think of just "getNode", but that may lead to confusion with the querySelector which is used to find nodes, but only returns the nodeId[s]).

**Comment by pavelfeldman** [comment]
Present DOM domain serves the 'inspection' goal, not scripting or
automation. If you have nodeId in hand, it means that all the node
information has already been sent to you in one of the responses or
notifications.

**Comment by mafredri** [comment]
@pavelfeldman to see if I'm understanding this correctly. Given that the `DOM.attributeModified` can return a `NodeId`, how would one look it up? Is it your intention that we listen to e.g. `DOM.childNodeInserted` (or `Removed`) to keep a local representation of the DOM, thus allowing us to look it up?

**Comment by pavelfeldman** [comment]
> Is it your intention that we listen to e.g. DOM.childNodeInserted (or Removed) to keep a local representation of the DOM

this is not my intention, but that's how DOM domain historically was designed. It was optimized for the inspection (devtools elements panel) and hence backend knows what nodes it has already sent to the client, notifies client of changes to these nodes, etc.

I'd like to introduce another DOM2 domain or mode of operation that would be more appropriate for automation / scripting. In that world, `nodeId` would not exist, unique `backendNodeId` would identify node and you would be able to fetch node info by its `backendNodeId`. That will take some time though. For now, caching all the info on nodes as they arrive is the only option.

**Comment by mafredri** [comment]
> this is not my intention, but that's how DOM domain historically was designed.

Sorry, saying it was *your intention* was perhaps a bad choice of words, I meant from the perspective of the protocol.

Anyway, thanks for the reply, it makes sense and definitely cleared it up for me. Also, looking forward to someday having DOM2, or the like 😄!

**Comment by rawbin-** [comment]
Is the DOM.resolveNode  what you want？ get the Node with the given nodeId
https://chromedevtools.github.io/devtools-protocol/tot/DOM/#method-resolveNode

**Comment by kensoh** [comment]
Adding on, @rbairwell it seems that you want the element instead of the ID, maybe to work on it. Like what Pavel said, the protocol was originally designed for debugging the browser. Although now with headless mode Chrome I think it is going to attract attention from test / web automation developers.

At the moment I use Runtime.evaluate and the following snippets to get the element whether via CSS selector or XPath selector - for CSS - `querySelector(selector).action/property` and for XPath - `document.evaluate(selector,document,null,XPathResult.ORDERED_NODE_SNAPSHOT_TYPE,null).snapshotItem(0).method/property`

I don't have expectation that more changes will come to meet the wishes of automation use cases, and I don't know Google/Chrome longterm goal in opening up headless mode, but being browser agnostic (Edge, Safari etc), it is going to shake-up the Selenium / WebdriverIO world.

**Comment by fjwgrace** [comment]
> - f
but  how to get the cssSelectors or Xpath via Chrome devTools Protocol?

**Comment by kensoh** [comment]
Not sure what is meant by how to 'get', because CSS selectors and XPath selectors are normally 'given' to Chrome to tell it which element to automate on. For me I use `Runtime.evaluate` and the JS code in my above comment to act on elements using CSS or XPath selectors accordingly.

**Comment by fjwgrace** [comment]
> Not sure what is meant by how to 'get', because CSS selectors and XPath selectors are normally 'given' to Chrome to tell it which element to automate on. For me I use `Runtime.evaluate` and the JS code in my above comment to act on elements using CSS or XPath selectors accordingly.

I want obtain the css selectrol by chrome devtools Protocol？

**Comment by TimvdLippe** [comment]
This repository is related to Chrome DevTools Protocol, but does not track issues regarding its definition or implementation. If you want to file an issue for the Chrome DevTools Protocol, please open an issue on https://crbug.com under [`component: Platform>DevTools>Platform`](https://bugs.chromium.org/p/chromium/issues/entry?components=Platform%3EDevTools%3EPlatform). Thanks in advance!

**Comment by s-janibekova** [comment]
> > * f
> >   but  how to get the cssSelectors or Xpath via Chrome devTools Protocol?

hello. Did you find the answer? I am also looking for the same question

**Comment by kensoh** [comment]
Maybe need to write logic to generate base on attributes present and test iteratively until something unique is found?

Copying @sanjayselectorshub (creator of SelectorsHub and ChroPath) -
on this interesting question on how to get XPath from a specified web element

**Comment by giuliohome** [comment]
> This repository is related to Chrome DevTools Protocol, but does not track issues regarding its definition or implementation. If you want to file an issue for the Chrome DevTools Protocol, please open an issue on https://crbug.com under [`component: Platform>DevTools>Platform`](https://bugs.chromium.org/p/chromium/issues/entry?components=Platform%3EDevTools%3EPlatform). Thanks in advance!

@TimvdLippe  Sorry, so you are confirm that current Chrome DevTools Protocol does not provide for finding an id, aren't you? 

>Present DOM domain serves the 'inspection' goal, not scripting or
automation

@pavelfeldman That's also quite surprising, when you consider that Selenium WebDriver is mainly used for Web Automation and Selenium WebDriver offers Selenium Locators. 

So, my question to you (@TimvdLippe @pavelfeldman etc...) boils down to asking whether Selenium Locators are not based on Chrome DevTools Protocol, whilst - on the opposite -  e.g. `sendKeys `is based the `Input.dispatchKeyEvent` of Chrome DevTools Protocol. 
In other words, does the `driver.findElement(By.id` actually translate into a `Runtime.evaluate` of Chrome DevTools Protocol, as we all are thinking (see also the [above comment](https://github.com/ChromeDevTools/devtools-protocol/issues/20#issuecomment-315536099))? Thank you in advance for this confirmation!
